File: mobile/src/main/java/uk/co/appsbystudio/geoshare/utils/dialog/ShareOptions.java
package uk.co.appsbystudio.geoshare.utils.dialog;

import android.app.AlertDialog;
import android.app.Dialog;
import android.app.DialogFragment;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.widget.Toast;

import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;

import uk.co.appsbystudio.geoshare.Application;
import uk.co.appsbystudio.geoshare.GPSTracking;
import uk.co.appsbystudio.geoshare.R;
import uk.co.appsbystudio.geoshare.utils.firebase.DatabaseLocations;
import uk.co.appsbystudio.geoshare.utils.firebase.FirebaseHelper;
import uk.co.appsbystudio.geoshare.utils.firebase.TrackingInfo;
import uk.co.appsbystudio.geoshare.utils.services.TrackingService;

public class ShareOptions extends DialogFragment {

    private SharedPreferences sharedPreferences;

    private String friendId;
    private String uid;

    private GPSTracking gpsTracking;

    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {

        Bundle args = getArguments();
        String name = args.getString("name");
        friendId = args.getString("friendId");
        uid = args.getString("uid");

        gpsTracking = new GPSTracking(getActivity());

        sharedPreferences = getActivity().getSharedPreferences("tracking", Context.MODE_PRIVATE);

        AlertDialog.Builder optionsMenu = new AlertDialog.Builder(getActivity(), R.style.DialogTheme);

        optionsMenu.setTitle("Share your location with " + name + "?").setMultiChoiceItems(R.array.shareLocationOptions, null, new DialogInterface.OnMultiChoiceClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which, boolean isChecked) {
                switch (which) {
                    case 0:
                        if (!isChecked) ((AlertDialog) dialog).getListView().setItemChecked(1, false);
                        break;
                    case 1:
                        if (isChecked) ((AlertDialog) dialog).getListView().setItemChecked(0, true);
                        break;
                }
            }
        }).setPositiveButton("Share", new OnPositiveButtonClick()).setNegativeButton("Cancel", null);

        return optionsMenu.create();
    }

    private class OnPositiveButtonClick implements DialogInterface.OnClickListener {
        @Override
        public void onClick(DialogInterface dialog, int i) {
            if (((AlertDialog) dialog).getListView().getCheckedItemPositions().get(0) && !((AlertDialog) dialog).getListView().getCheckedItemPositions().get(1)) {
                DatabaseLocations databaseLocations = new DatabaseLocations(gpsTracking.getLongitude(), gpsTracking.getLatitude(), System.currentTimeMillis());

                DatabaseReference databaseReference = FirebaseDatabase.getInstance().getReference();
                if (uid != null && friendId != null) {
                    databaseReference.child("current_location").child(friendId).child(uid).setValue(databaseLocations)
                            .addOnSuccessListener(new OnSuccessListener<Void>() {
                                @Override
                                public void onSuccess(Void aVoid) {
                                    Toast.makeText(Application.getContext(), "Success", Toast.LENGTH_LONG).show();
                                }
                            })
                            .addOnFailureListener(new OnFailureListener() {
                                @Override
                                public void onFailure(@NonNull Exception e) {
                                    Toast.makeText(Application.getContext(), "Failed", Toast.LENGTH_LONG).show();
                                }
                            });
                    sharedPreferences.edit().putBoolean(friendId, false).apply();
                }
            } else if (((AlertDialog) dialog).getListView().getCheckedItemPositions().get(0) && ((AlertDialog) dialog).getListView().getCheckedItemPositions().get(1)) {
                DatabaseReference databaseReference = FirebaseDatabase.getInstance().getReference();
                if (uid != null && friendId != null) {
                    TrackingInfo trackingInfo = new TrackingInfo(true, System.currentTimeMillis());
                    databaseReference.child(FirebaseHelper.TRACKING).child(friendId).child("tracking").child(uid).setValue(trackingInfo)
                            .addOnSuccessListener(new OnSuccessListener<Void>() {
                                @Override
                                public void onSuccess(Void aVoid) {
                                    sharedPreferences.edit().putBoolean(friendId, true).apply();
                                    if (!TrackingService.isRunning) {
                                        Intent trackingService = new Intent(Application.getContext(), TrackingService.class);
                                        Application.getContext().startService(trackingService);
                                    }
                                    Toast.makeText(Application.getContext(), "Success", Toast.LENGTH_LONG).show();
                                }
                            })
                            .addOnFailureListener(new OnFailureListener() {
                                @Override
                                public void onFailure(@NonNull Exception e) {
                                    Toast.makeText(Application.getContext(), "Failed", Toast.LENGTH_LONG).show();
                                }
                            });

                    DatabaseLocations databaseLocations = new DatabaseLocations(gpsTracking.getLongitude(), gpsTracking.getLatitude(), System.currentTimeMillis());

                    databaseReference.child(FirebaseHelper.TRACKING).child(uid).child("location").setValue(databaseLocations);

                    databaseReference.child("current_location").child(friendId).child(uid).removeValue();
                }
            }
        }
    }
}

File: mobile/src/main/java/uk/co/appsbystudio/geoshare/utils/dialog/ShareOptions.kt
package uk.co.appsbystudio.geoshare.utils.dialog

import android.app.AlertDialog
import android.app.Dialog
import android.app.DialogFragment
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.widget.Toast
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.FirebaseDatabase
import uk.co.appsbystudio.geoshare.R
import uk.co.appsbystudio.geoshare.utils.GPSTracking
import uk.co.appsbystudio.geoshare.utils.firebase.DatabaseLocations
import uk.co.appsbystudio.geoshare.utils.firebase.FirebaseHelper
import uk.co.appsbystudio.geoshare.utils.firebase.TrackingInfo
import uk.co.appsbystudio.geoshare.utils.services.TrackingService
import uk.co.appsbystudio.geoshare.utils.ui.notifications.NewRequestNotification
import java.text.SimpleDateFormat
import java.util.*

class ShareOptions : DialogFragment() {

    private var mContext: Context? = null

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        this.mContext = context
    }
    
    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {

        val args = arguments
        val name = args.getString("name")
        val friendId = args.getString("friendId")
        val uid = FirebaseAuth.getInstance().currentUser?.uid

        NewRequestNotification.cancel(mContext, friendId)

        val gpsTracking = GPSTracking(mContext)

        val trackingPreferences = mContext?.getSharedPreferences("tracking", Context.MODE_PRIVATE)
        val timerPreferences = mContext?.getSharedPreferences("tracking_time", Context.MODE_PRIVATE)

        val optionsMenu = AlertDialog.Builder(mContext, R.style.DialogTheme)

        optionsMenu.setTitle("Share your location with $name?").setMultiChoiceItems(R.array.shareLocationOptions, null) { dialog, which, isChecked ->
            when (which) {
                0 -> if (!isChecked) (dialog as AlertDialog).listView.setItemChecked(1, false)
                1 -> if (isChecked) (dialog as AlertDialog).listView.setItemChecked(0, true)
            }
        }.setPositiveButton("Share", null).setNegativeButton("Cancel") { _, _ -> dismiss()}

        val dialog = optionsMenu.create()

        dialog.show()

        dialog.getButton(AlertDialog.BUTTON_POSITIVE).setOnClickListener {
            when {
                dialog.listView.checkedItemPositions.get(0) && !dialog.listView.checkedItemPositions.get(1) -> {
                    val lastShare = if (timerPreferences != null) System.currentTimeMillis() - timerPreferences.getLong(friendId, 0) else 0
                    if (lastShare > 300000) { //5 mins in millis
                        val databaseLocations = DatabaseLocations(gpsTracking.getLongitude(), gpsTracking.getLatitude(), System.currentTimeMillis())

                        val databaseReference = FirebaseDatabase.getInstance().reference
                        if (uid != null && friendId != null) {
                            databaseReference.child("current_location").child(friendId).child(uid).setValue(databaseLocations)
                                    .addOnSuccessListener {
                                        timerPreferences?.edit()?.putLong(friendId, System.currentTimeMillis())?.apply()
                                    }.addOnFailureListener {
                                        Toast.makeText(mContext, "Failed", Toast.LENGTH_LONG).show()
                                    }
                            trackingPreferences?.edit()?.putBoolean(friendId, false)?.apply()
                        }
                    } else {
                        val timeLeft = (300000 - lastShare)
                        val format = SimpleDateFormat("mm:ss", Locale.getDefault())
                        Toast.makeText(mContext, "Please wait another ${format.format(timeLeft)}", Toast.LENGTH_LONG).show()
                    }
                }
                dialog.listView.checkedItemPositions.get(0) && dialog.listView.checkedItemPositions.get(1) -> {
                    val databaseReference = FirebaseDatabase.getInstance().reference
                    if (uid != null && friendId != null) {
                        val trackingInfo = TrackingInfo(true, System.currentTimeMillis())
                        databaseReference.child(FirebaseHelper.TRACKING).child(friendId).child("tracking").child(uid).setValue(trackingInfo)
                                .addOnSuccessListener {
                                    trackingPreferences?.edit()?.putBoolean(friendId, true)?.apply()
                                    if (!TrackingService.isRunning) {
                                        val trackingService = Intent(mContext, TrackingService::class.java)
                                        mContext?.startService(trackingService)
                                    }
                                }.addOnFailureListener {
                                    Toast.makeText(mContext, "Failed", Toast.LENGTH_LONG).show()
                                }

                        val databaseLocations = DatabaseLocations(gpsTracking.getLongitude(), gpsTracking.getLatitude(), System.currentTimeMillis())

                        databaseReference.child(FirebaseHelper.TRACKING).child(uid).child("location").setValue(databaseLocations)

                        databaseReference.child("current_location").child(friendId).child(uid).removeValue()
                    }
                }
            }
            dismiss()
        }

        return dialog
    }
}
